pkg/controller/handlers/mcpcatalog/mcpcatalog.go

```diff
@@ -0,0 +1,467 @@
+	"github.com/obot-platform/nah/pkg/router"
+	"github.com/obot-platform/obot/apiclient/types"
+	"github.com/obot-platform/obot/pkg/controller/handlers/toolreference"
+	"github.com/obot-platform/obot/pkg/controller/handlers/usercatalogauthorization"
+	gclient "github.com/obot-platform/obot/pkg/gateway/client"
+	v1 "github.com/obot-platform/obot/pkg/storage/apis/obot.obot.ai/v1"
+	"github.com/obot-platform/obot/pkg/system"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Handler struct {
+	allowedDockerImageRepos []string
+	defaultCatalogPath      string
+	gatewayClient           *gclient.Client
+}
+
+func New(allowedDockerImageRepos []string, defaultCatalogPath string, gatewayClient *gclient.Client) *Handler {
+	return &Handler{
+		allowedDockerImageRepos: allowedDockerImageRepos,
+		defaultCatalogPath:      defaultCatalogPath,
+		gatewayClient:           gatewayClient,
+	}
+}
+
+func (h *Handler) Sync(req router.Request, resp router.Response) error {
+	mcpCatalog := req.Object.(*v1.MCPCatalog)
+	toAdd := make([]client.Object, 0)
+
+	for _, sourceURL := range mcpCatalog.Spec.SourceURLs {
+		objs, err := h.readMCPCatalog(mcpCatalog.Name, sourceURL)
+		if err != nil {
+			return fmt.Errorf("failed to read catalog %s: %w", sourceURL, err)
+		}
+
+		toAdd = append(toAdd, objs...)
+	}
+
+	mcpCatalog.Status.LastSyncTime = metav1.Now()
+	if err := req.Client.Status().Update(req.Ctx, mcpCatalog); err != nil {
+		return fmt.Errorf("failed to update catalog status: %w", err)
+	}
+
+	// We want to refresh this every hour.
+	// TODO: make this configurable.
+	resp.RetryAfter(time.Hour)
+
+	if len(toAdd) == 0 {
+		// If there's nothing to add, then that means that there are no entries from the source URLs.
+		// Now we will delete all non-editable entries, since those came from a remote source in the past,
+		// and thus no longer exist. The apply function will not remove them if toAdd is empty, so we need todo it manually.
+
+		var entries v1.MCPServerCatalogEntryList
+		if err := req.Client.List(req.Ctx, &entries, client.InNamespace(system.DefaultNamespace), client.MatchingFields{
+			"spec.mcpCatalogName": mcpCatalog.Name,
+		}); err != nil {
+			return fmt.Errorf("failed to list entries: %w", err)
+		}
+
+		for _, entry := range entries.Items {
+			if !entry.Spec.Editable {
+				if err := req.Client.Delete(req.Ctx, &entry); err != nil {
+					return fmt.Errorf("failed to delete entry %s: %w", entry.Name, err)
+				}
+			}
+		}
+
+		return nil
+	}
+
+	// I know we don't want to do apply anymore. But we were doing it before in a different place.
+	// Now we're doing it here. It's not important enough to change right now.
+	return apply.New(req.Client).WithOwnerSubContext(mcpCatalog.Name).Apply(req.Ctx, mcpCatalog, toAdd...)
+}
+
+func (h *Handler) readMCPCatalog(catalogName, sourceURL string) ([]client.Object, error) {
```

Comment:
Virtually all of the code from this function all the way down to the start of the next proper controller route was copied over from the ToolReference handler, which used to be responsible for setting up the MCPServerCatalogEntries. Very little of this is new logic.